### ISSUE TITLE
components are appended after the gtk widgets

### ISSUE THREAD

**Issue by maruns** [AaronErhardt/relm4#20]
I'm learning rust so please be lenient with me...

When appending components to the gtk::Box together with other gtk::widgets, components are added at the end.

```
#[widget]
impl Widgets<MainWindowModel, ()> for MainWindow {
    view! {
        gtk::ApplicationWindow {
            set_title: Some("main window"),
            set_default_width: 600,
            set_default_height: 300,

            set_child = Some(&gtk::Box) {
                set_margin_all: 0,
                set_orientation: gtk::Orientation::Horizontal,
                
                // append = &gtk::Box {
                    // set_orientation: gtk::Orientation::Vertical,

                    append: component!(components.profile_view.root_widget()),
                // },
                append = &gtk::Paned {
                    set_orientation: gtk::Orientation::Horizontal,
                    set_position: 250,
                    set_shrink_start_child: false,
                    set_shrink_end_child: false,
                    set_start_child = &gtk::Button{
                        set_label: "sources",
                        set_hexpand: false,

                    },
                    set_end_child = &gtk::Paned {
                        set_orientation: gtk::Orientation::Horizontal,
                        set_position: 250,
                        set_shrink_start_child: false,
                        set_shrink_end_child: false,
                        set_start_child = &gtk::Button {
                            set_label: "items",
                        },
                        set_end_child: component!(components.item_view.root_widget()),

                    }

                },
                
                
            }
        }
    }
}
```

Leads to the main window looking like:

![main_window_orig](https://user-images.githubusercontent.com/1474085/132122203-89fded74-b497-48c7-a7b0-f63200eb318c.png)

While I would expect

![component_in_a_box](https://user-images.githubusercontent.com/1474085/132122247-49cc9144-0566-4ea2-9d68-548bb16b63fb.png)

Which can be achieved by uncommenting the box wrapping around the `append: component!(components.profile_view.root_widget()),`.

My profile view  widget looks like 

```
#[widget(pub)]
impl<ParentModel> Widgets<ProfileViewModel, ParentModel> for ProfileView 
    where
        ParentModel: ProfileViewParent,
        ParentModel::Widgets: ProfileViewParentWidgets,
{
    
    view!{
        gtk::Box{
            set_orientation: gtk::Orientation::Vertical,
            set_margin_all: 5,
            set_spacing: 5,
            factory!(model.profiles),
        }
    }
}
```

Since root of the `ProfileView` is a vertical gtk::Box I would expect wrapping it with another vertical box to be idempotent in terms of view (ignoring margins, paddings, borders, etc...). 

-----

Factory for profile buttons:

```
impl FactoryPrototype for ProfileViewItem {
    type Factory = FactoryVecDeque<Self>;
    type Widgets = ProfileViewItemWidgets;
    type Root = gtk::Button;
    type View = gtk::Box;
    type Msg = ProfileViewMsg;

    fn position(&self, _index: &Rc<DynamicIndex>) {}

    fn update(&self, _index: &Rc<DynamicIndex>, _widgets: &ProfileViewItemWidgets) {}

    fn get_root(widgets: &ProfileViewItemWidgets) -> &gtk::Button {
        &widgets.profile_button
    }

    fn generate(
        &self,
        _index: &Rc<DynamicIndex>,
        _sender: Sender<ProfileViewMsg>
    ) -> ProfileViewItemWidgets {
        let mut graphemes = UnicodeSegmentation::graphemes(
            self.profile.name.as_str(),
            true,
        );
        let c = match graphemes.next() {
            Some(v) => v,
            None => "?"
        };

        let profile_button = gtk::Button::with_label(c);

        profile_button.add_css_class("flat");

        ProfileViewItemWidgets {
            profile_button,
        }
    }
}
```

**Comment by AaronErhardt** [comment]
That's a general problem in Relm4 that can't be really fixed. The thing is that the components are initialized after their parent so adding components to your widgets must happen later (all other things should still be in order). 

Fortunately, you can work around this behavior by using the [prepend method](https://gtk-rs.org/gtk4-rs/git/docs/gtk4/prelude/trait.BoxExt.html#tymethod.prepend) on the box. Simply replace append with prepend. That tells GTK to add the widget to the start so you don't have any issues with the order anymore :)

**Comment by maruns** [comment]
Prepend works for me.

Only thing I would like to see in this case is big red warning in a book in component's chapter.

**Comment by AaronErhardt** [comment]
I've thought about this issue a bit and actually found a way to solve this. If everything goes well, I'll release version 0.2 soon with a fixed version of the macro. Until then I'll keep this issue open.

**Comment by AaronErhardt** [comment]
Fixed in Relm4 v0.2
